### PR TITLE
Allow user to specify custom typeshed directory

### PR DESCRIPTION
**Pull request by ddfisher**
This avoids doing any sort of cleanup of `default_data_dir` to be as safe of a change as possible.

Triggers #1587 and therefore blocks on it.

### CODE REVIEW

File: mypy/main.py [hunk 1]
```diff
@@ -160,6 +162,8 @@ def parse_version(v):
                         help="an anti-pattern")
     parser.add_argument('--stats', action='store_true', help="dump stats")
     parser.add_argument('--inferstats', action='store_true', help="dump type inference stats")
+    parser.add_argument('--custom-typeshed', metavar='PATH',
+                        help="use the custom typeshed module at PATH")
```
Comment (JukkaL) [inline mypy/main.py]: I wonder if the term 'typeshed module' is clear enough, as the term 'module' is overloaded? Maybe use 'typeshed repository'?

**Comment by JukkaL** [comment]
Looks good, other the note above about the help message.

File: mypy/main.py (hunk 1 above)
Comment (ddfisher) [inline mypy/main.py]: What about "typeshed package"?

**Comment by gvanrossum** [comment]
It's neither a package nor a module -- it's a directory though.

Maybe now that you've redone the command line options you can rebase, rework, and resubmit this? I still think it's a great idea. (Maybe also add to the docs? There's an extensive chapter on the command line.)

**Comment by ddfisher** [comment]
I think this is blocking on #1587 now.

File: mypy/build.py [hunk 2]
```diff
@@ -125,12 +125,20 @@ def build(sources: List[BuildSource],
...
     """
 
-    data_dir = default_data_dir(bin_dir)
+    custom_typeshed_name = None
+    if options.custom_typeshed_dir:
+        custom_data_dir, custom_typeshed_name = os.path.split(
+            options.custom_typeshed_dir.rstrip(os.sep))
+        data_dir = custom_data_dir
+    else:
+        data_dir = default_data_dir(bin_dir)
```
Comment (gvanrossum) [inline mypy/build.py]: Can you work around #1587 by setting `custom_typeshed_name = None` here?

Comment (ddfisher) [inline mypy/build.py]: Yep, that fixes it!

**Comment by ddfisher** [comment]
Updated!  Should be mergeable now.

File: mypy/main.py [hunk 3]
```diff
@@ -188,6 +188,8 @@ def process_options(args: List[str],
                         help="dump type inference stats")
     parser.add_argument('--custom-typing', metavar='MODULE', dest='custom_typing_module',
                         help="use a custom typing module")
+    parser.add_argument('--custom-typeshed-dir', metavar='DIR',
+                        help="use the custom typeshed in DIR")
```
Comment (gvanrossum) [inline mypy/main.py]: I think Jukka had suggested to change this to "custom typeshed *package* in DIR" ?

**Comment by gvanrossum** [comment]
PS can you try not to squash all local commits?

File: docs/source/command_line.rst [hunk 4]
```diff
@@ -326,6 +326,12 @@ Here are some more useful flags:
   package, the behavior enabled by this flag is often more
   convenient.)
 
+- ``--custom-typeshed-dir DIR`` specifies the directory where mypy looks for
```
Comment (ddfisher) [inline docs/source/command_line.rst]: 👍

File: mypy/build.py [hunk 5]
```diff
@@ -127,12 +127,20 @@ def build(sources: List[BuildSource],
         directories; if omitted, use '.' as the data directory
     """
 
-    data_dir = default_data_dir(bin_dir)
+    if options.custom_typeshed_dir:
+        custom_data_dir, custom_typeshed_name = os.path.split(
+            options.custom_typeshed_dir.rstrip(os.sep))
+        data_dir = custom_data_dir
+    else:
+        custom_typeshed_name = None
```
Comment (ddfisher) [inline mypy/build.py]: Yeah, that's nicer.  Changed.

File: mypy/build.py [hunk 6]
```diff
@@ -249,10 +257,13 @@ def mypy_path() -> List[str]:
...
-def default_lib_path(data_dir: str, pyversion: Tuple[int, int]) -> List[str]:
+def default_lib_path(data_dir: str,
+                     pyversion: Tuple[int, int],
+                     custom_typeshed_name: str = None) -> List[str]:
     """Return default standard library search paths."""
     # IDEA: Make this more portable.
     path = []  # type: List[str]
+    typeshed_name = custom_typeshed_name or 'typeshed'
 
     auto = os.path.join(data_dir, 'stubs-auto')
```
Comment (ddfisher) [inline mypy/build.py]: What's going on here?  What is `stubs-auto`?  (The commit that introduces it doesn't explain at all...)

File: mypy/main.py (hunk 3 above)
Comment (ddfisher) [inline mypy/main.py]: Looking through the old comments, we noted that typeshed wasn't really a package -- that's why I left that out here.

File: mypy/build.py [hunk 7]
```diff
@@ -266,7 +277,7 @@ def default_lib_path(data_dir: str, pyversion: Tuple[int, int]) -> List[str]:
     # (Note that 3.1 and 3.0 aren't really supported, but we don't care.)
     for v in versions + [str(pyversion[0]), '2and3']:
         for lib_type in ['stdlib', 'third_party']:
-            stubdir = os.path.join(data_dir, 'typeshed', lib_type, v)
+            stubdir = os.path.join(data_dir, typeshed_name, lib_type, v)
```
Comment (ddfisher) [inline mypy/build.py]: Refactored.

File: mypy/build.py (hunk 5 above)
Comment (ddfisher) [inline mypy/build.py]: Actually, this got changed in a different way from the refactor to fix the other issues.

**Comment by ddfisher** [comment]
@gvanrossum can you take another look at this?

**Comment by ddfisher** [comment]
🎉

**Pull request merged**